test/unit/math/prim/prob/hmm_marginal_test.cpp

```diff
@@ -0,0 +1,243 @@
+    for (int i = 0; i < rho.size() - 1; i++)
+      rho_unconstrained_[i] = rho(i);
+    rho_unconstrained = rho_unconstrained_;
+
+    Gamma_unconstrained = Gamma.block(0, 0, n_states, n_states - 1);
+  }
+
+  int n_states, n_transitions;
+  double abs_mu, sigma, p1_init, gamma1, gamma2;
+
+  Eigen::VectorXd rho;
+  Eigen::MatrixXd Gamma;
+  Eigen::VectorXd obs_data;
+  Eigen::MatrixXd log_omegas;
+  Eigen::MatrixXd log_omegas_zero;
+
+  // Construct "unconstrained" versions of rho and Gamma, without
+  // the final element which can be determnied using the fact
+  // the columns sum to 1. This allows us to do finite diff tests,
+  // without violating the simplex constraint of rho and Gamma.
+  std::vector<double> rho_unconstrained;
+  Eigen::MatrixXd Gamma_unconstrained;
+  stan::test::ad_tolerances tols;
+};
+
+TEST_F(hmm_marginal_lpdf_test, ten_transitions) {
+  using stan::math::hmm_marginal_lpdf;
+
+  EXPECT_FLOAT_EQ(-18.37417, hmm_marginal_lpdf(log_omegas, Gamma, rho));
+
+  // Differentiation tests
+  auto hmm_functor = [](const auto& log_omegas, const auto& Gamma_unconstrained,
+                        const auto& rho_unconstrained) {
+    return hmm_marginal_test_wrapper(log_omegas, Gamma_unconstrained,
+                                     rho_unconstrained);
+  };
+
+  stan::test::expect_ad(tols, hmm_functor, log_omegas, Gamma_unconstrained,
+                        rho_unconstrained);
+}
+
+TEST_F(hmm_marginal_lpdf_test, zero_transitions) {
+  using stan::math::hmm_marginal_lpdf;
+
+  EXPECT_FLOAT_EQ(-1.520827, hmm_marginal_lpdf(log_omegas_zero, Gamma, rho));
+
+  // Differentiation tests
+  auto hmm_functor = [](const auto& log_omegas, const auto& Gamma_unconstrained,
+                        const auto& rho_unconstrained) {
+    return hmm_marginal_test_wrapper(log_omegas, Gamma_unconstrained,
+                                     rho_unconstrained);
+  };
+
+  stan::test::expect_ad(tols, hmm_functor, log_omegas_zero, Gamma_unconstrained,
+                        rho_unconstrained);
+}
+
+TEST(hmm_marginal_lpdf, one_state) {
+  using stan::math::hmm_marginal_lpdf;
+  int n_states = 1, p1_init = 1, gamma1 = 1, n_transitions = 10, abs_mu = 1,
+      sigma = 1;
+  Eigen::VectorXd rho(n_states);
+  rho << p1_init;
+  Eigen::MatrixXd Gamma(n_states, n_states);
+  Gamma << gamma1;
+  Eigen::VectorXd obs_data(n_transitions + 1);
+  obs_data << -0.9692032, 1.6367754, 1.0339449, 0.9798393, 0.4829358, 2.7508704,
+      0.3122448, 1.8316583, 1.6327319, 1.2097332, 0.4087620;
+  Eigen::MatrixXd log_omegas(n_states, n_transitions + 1);
+  for (int n = 0; n < n_transitions + 1; n++)
+    log_omegas.col(n)[0] = state_lpdf(obs_data[n], abs_mu, sigma, 0);
+
+  EXPECT_FLOAT_EQ(-14.89646, hmm_marginal_lpdf(log_omegas, Gamma, rho));
+
+  // Differentiation tests
+  // In the case where we have one state, Gamma and rho
+  // are fixed (i.e = 1)
+  auto hmm_functor = [](const auto& log_omegas) {
+    Eigen::MatrixXd Gamma(1, 1);
```

Comment:
This can be a one-liner (spread over two lines):

```
  return marginal_lpdf(log_omegas, Eigen::MatrixXd::Constant(1, 1, 1),
                       Eigen::VectorXd::Constant(1, 1));
```

Reply:
To me, it's a bit less clear to read, and you also get (petty) error message:
```
./stan/math/prim/prob/hmm_marginal_lpdf.hpp:72:13: note: candidate template
      ignored: could not match 'Matrix' against 'CwiseNullaryOp'
```
If you're ok with it, I'll leave it as is.

Reply:
OK.